### PR TITLE
feat(frontend): improved UX for missing vulnerabilities

### PR DESCRIPTION
**Pull request by andrewpollock**
This commit modifies the 404 flow so that:
- when a request for a non-existent vulnerability is received, the 404 page reflects the nature of the failed request
- when it is known that the vulnerability failed to import, this is explicitly surfaced

A new FAQ entry discussing this is also added.

Additionally, the 404 generation was tidied up to make it easier to understand where the 404 comes from. There were many unreachable paths that called `abort()`, which made the code unnecessarily unreadable.

It also addresses requests for `/favicon.ico` generating 404's from the redirector by adding an explicit handler for it.

127 0 0 1_8000_USN-2001-1

127 0 0 1_8000_foo_bar_baz

127 0 0 1_8000_list_q=USN-2001-1 ecosystem=

Part of #2190

### CODE REVIEW

File: gcp/appengine/frontend_handlers.py [hunk 1]
```diff
@@ -269,19 +285,17 @@ def vulnerability_json_redirector(potential_vuln_id):
   https://api.osv.dev/v1/vulns/VULN-ID.
   """
   if not _VALID_VULN_ID.match(potential_vuln_id):
-    abort(404)
+    abort(400)
```
[the remark this answers is not in the document]
Comment (andrewpollock) [reply gcp/appengine/frontend_handlers.py:288]: I think the 400 page looks the same as our 500 page (and is something we need to address). There's already 400's emitted elsewhere, so it's not a new phenomenon, and you have to go out of your way to trigger this one.

File: gcp/appengine/frontend3/src/templates/404.html [hunk 2]
```diff
@@ -12,6 +12,27 @@
...
 
 {% block content %}
+{% if vuln_id %}
+<div class="mdc-layout-grid not-found-page">
+    <div class="mdc-layout-grid__inner">
+        <div class="mdc-layout-grid__cell--span-12 text-info">
+            <h2 class="heading">{{ vuln_id }} Not Found!</h2>
+            <p class="description">
+            {% if has_importfindings %}
+            The record has not been successfully imported from its source<br/>
```
Comment (oliverchang) [inline gcp/appengine/frontend3/src/templates/404.html:22]: Do we plan to surface the actual errors? Or point to where a source provider can see what went wrong?

**Review by oliverchang — APPROVED** [review]

File: gcp/appengine/frontend_handlers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewpollock) [reply gcp/appengine/frontend_handlers.py:288]: > Huh isn't this triggered if you went to osv.dev/CVE_2024_1234 or something similar

No:

https://github.com/google/osv.dev/blob/master/gcp/appengine/frontend_handlers.py#L50-L53

is there to placate GitHub's code scanning whinging about using unsanitized user input in redirects. We'd have to bake in the [same regex that is used in the JSON Schema validation](https://github.com/ossf/osv-schema/blob/main/validation/schema.json#L349-L354) (with the attendant maintenance burden) to more rigorously validate it to catch examples like this.

File: gcp/appengine/frontend3/src/templates/404.html (hunk 2 above)
Comment (andrewpollock) [reply gcp/appengine/frontend3/src/templates/404.html:22]: > Do we plan to surface the actual errors? Or point to where a source provider can see what went wrong?

Yes, that's https://github.com/google/osv.dev/issues/2189

**Pull request merged**